### PR TITLE
Refactoring Dockerfile

### PR DESCRIPTION
**Pull request by foxdalas**
* Migrate to latest Ubuntu LTS (22.04)
* Build inside docker (buildx support)

### CODE REVIEW

File: Dockerfile [hunk 1]
```diff
@@ -15,16 +15,25 @@
...
 RUN apt update && apt install -y libsnappy-dev
 WORKDIR /kvrocks
 
 RUN mkdir /data
-COPY ./build/bin/kvrocks ./bin/
+COPY --from=build /kvrocks/src/kvrocks ./bin/
+
 COPY ./kvrocks.conf  ./conf/
 RUN sed  -i -e 's|dir /tmp/kvrocks|dir /var/lib/kvrocks|g' ./conf/kvrocks.conf
-VOLUME /var/lib/kvrocks
```
Comment (git-hulk) [inline Dockerfile:27]: @foxdalas Maybe we should keep this volume, some users would like to volume the host configuration file into docker.

**Comment by git-hulk** [comment]
@foxdalas Thanks for your contribution. No worry about the failure of CI tests, will find the root cause and fix it later.

**Comment by foxdalas** [comment]
@[git-hulk](https://github.com/git-hulk) Thanks. Maybe do you need ci for docker with many platforms?

File: Dockerfile (hunk 1 above)
Comment (foxdalas) [reply Dockerfile:27]: Yes sure

**Comment by git-hulk** [comment]
Yes, we need it. I have a try at my side but failed, seems `jemalloc` didn't work with latest version g++11. Does it work on your side?

**Comment by foxdalas** [comment]
@git-hulk I built current head on my laptop. You can pull image `foxdalas/kvrocks:latest` `amd64/arm64.`
Everything works in AWS EKS.

**Comment by git-hulk** [comment]
> @git-hulk I built current head on my laptop. You can pull image `foxdalas/kvrocks:latest` `amd64/arm64.` Everything works in AWS EKS.

Thanks @foxdalas, are you using this Dockerfile, I got the below error when compiling.

```shell
include/jemalloc/internal/rtree.h:109:33: error: variably modified 'root' at file scope
  109 |         rtree_leaf_elm_t        root[1U << (RTREE_NSB/RTREE_HEIGHT)];
```

**Comment by foxdalas** [comment]
> > @git-hulk I built current head on my laptop. You can pull image `foxdalas/kvrocks:latest` `amd64/arm64.` Everything works in AWS EKS.
> 
> Thanks @foxdalas, are you using this Dockerfile, I got the below error when compiling.
> 
> ```shell
> include/jemalloc/internal/rtree.h:109:33: error: variably modified 'root' at file scope
>   109 |         rtree_leaf_elm_t        root[1U << (RTREE_NSB/RTREE_HEIGHT)];
> ```

It's strange. I will check it tomorrow. Thanks.

**Comment by foxdalas** [comment]
@git-hulk please approve buildx actions.

**Comment by git-hulk** [comment]
Sorry that I also don't have the permission now. Maybe @tisonkun @Hexiaoqiao can help to approve it.

**Comment by foxdalas** [comment]
> Sorry that I also don't have the permission now. Maybe @tisonkun @Hexiaoqiao can help to approve it.

Thanks for support.

**Comment by foxdalas** [comment]
@tisonkun @Hexiaoqiao approve action please

**Comment by foxdalas** [comment]
@tisonkun @Hexiaoqiao one more run please

**Comment by foxdalas** [comment]
@git-hulk all checks have been passed. Please check your environment everything forks on ubuntu 22.04 in docker.

**Comment by git-hulk** [comment]
Cool, thanks. Will have a try later.

**Comment by tisonkun** [comment]
@foxdalas @git-hulk please note that docker workflow takes about 197 minutes to complete. We may not run such workflow for every PR every activities.

**Comment by foxdalas** [comment]
> @foxdalas @git-hulk please note that docker workflow takes about 197 minutes to complete. We may not run such workflow for every PR every activities.

It's qemu 😢 I think i will find workaround

**Comment by foxdalas** [comment]
@tisonkun I can add cross-compiling to project or we can build docker image on tag push. 
What do you think about release action with docker build and push ?

**Comment by git-hulk** [comment]
@foxdalas We have built and pushed the docker image when releasing the new tag. 

Can see: https://github.com/apache/incubator-kvrocks/blob/unstable/.github/workflows/release.yaml#L80

**Comment by git-hulk** [comment]
@foxdalas We have built and pushed the docker image: [release.yaml#L80](https://github.com/apache/incubator-kvrocks/blob/unstable/.github/workflows/release.yaml#L80). I think we can add the cross compiling on the daily CI instead of every PR, HDYT?

**Comment by foxdalas** [comment]
> @foxdalas We have built and pushed the docker image: [release.yaml#L80](https://github.com/apache/incubator-kvrocks/blob/unstable/.github/workflows/release.yaml#L80). I think we can add the cross compiling on the daily CI instead of every PR, HDYT?

Do you mean docker buildx cross compiling? It's easy

**Comment by git-hulk** [comment]
Yes, as @tisonkun mentioned, the new docker job needs 197 minutes to complete, it's too slowly and unnecessary for most PR. So it would be better to add to Daily CI.

**Comment by foxdalas** [comment]
@git-hulk @tisonkun review please

File: .github/workflows/kvrocks.yaml [hunk 2]
```diff
@@ -93,3 +93,28 @@ jobs:
 
       - name: Build
         run: make -j4
+
```
Comment (git-hulk) [inline .github/workflows/kvrocks.yaml]: should remove this part?

**Review by git-hulk — APPROVED** [review]

**Review by vongosling — APPROVED** [review]

**Comment by ShooterIT** [comment]
did you run daily CR? works well? how long did it take? since currently we only run  `kvrocks.yaml` CI when submitting a PR.

**Comment by foxdalas** [comment]
> did you run daily CR? works well? how long did it take? since currently we only run `kvrocks.yaml` CI when submitting a PR.

197 minutes to complete for daily ci. Everything works on Graviton 2 in AWS.

**Comment by git-hulk** [comment]
@foxdalas Can you rollback the cross compile in `release.yaml`? For Daily CI, 197 minutes was ok to me, but it's too slowly to release. My point was we want to guarantee the code can compile on different platforms first.

**Comment by foxdalas** [comment]
> @foxdalas Can you rollback the cross compile in `release.yaml`? For Daily CI, 197 minutes was ok to me, but it's too slowly to release. My point was we want to guarantee the code can compile on different platforms first.

Yes,sure. 
Do you have another way for arm64 release?

**Comment by ShooterIT** [comment]
@git-hulk Maybe i didn't describe clearly. I think `release.yaml` is OK since it runs in the background.  But daily Ci takes much time.

Let's keep current solution. We can optimize if there is a real problem.

**Review by ShooterIT — APPROVED** [review]

**Comment by git-hulk** [comment]
OK, I'm ok to go head with the current way. We can improve this before GitHub Actions supporting the arm64 runner.

**Pull request merged**